src/System.IO.FileSystem.Watcher/src/System/IO/FileSystemWatcher.OSX.cs

```diff
@@ -191,224 +204,233 @@ private void CreateStreamAndStartWatcher()
+                    // Try to get the Watcher to raise the error event; if we can't do that, just silently exist since the watcher is gone anyway
+                    FileSystemWatcher watcher;
+                    if (_weakWatcher.TryGetTarget(out watcher))
+                    {
+                        // An error occurred while trying to start the run loop so fail out
+                        watcher.OnError(new ErrorEventArgs(new IOException(SR.EventStream_FailedToStart, Marshal.GetLastWin32Error())));
+                    }
+                }
 
-            // Since renames come in pairs, when we find the first we need to search for the next one. Once we find it, we'll add it to this
-            // list so when the for-loop comes across it, we'll skip it since it's already been processed as part of the original of the pair.
-            List<FSEventStreamEventId> handledRenameEvents = null;
+                // Always unschedule the RunLoop before cleaning up
+                Interop.EventStream.FSEventStreamUnscheduleFromRunLoop(_eventStream, _watcherRunLoop, Interop.RunLoop.kCFRunLoopDefaultMode);
+            }
 
-            for (long i = 0; i < numEvents.ToInt32(); i++)
+            private void FileSystemEventCallback( 
+                FSEventStreamRef streamRef, 
+                IntPtr clientCallBackInfo, 
+                size_t numEvents, 
+                String[] eventPaths, 
+                [MarshalAs(UnmanagedType.LPArray, SizeParamIndex = 2)]
+                Interop.EventStream.FSEventStreamEventFlags[] eventFlags,
+                [MarshalAs(UnmanagedType.LPArray, SizeParamIndex = 2)]
+                FSEventStreamEventId[] eventIds)
             {
-                // We need to special case the root path for pattern matching, so cache the path this way in case it is the root
-                // and use it for pattern matching. We'll go back to the eventPaths[i] to get the actual path for notifications
-                string path = eventPaths[i];
-                if (path.Equals(_fullDirectory, StringComparison.OrdinalIgnoreCase))
+                Debug.Assert((numEvents.ToInt32() == eventPaths.Length) && (numEvents.ToInt32() == eventFlags.Length) && (numEvents.ToInt32() == eventIds.Length));
+
+                // Try to get the actual watcher from our weak reference.  We maintain a weak reference most of the time
+                // so as to avoid a rooted cycle that would prevent our processing loop from ever ending
+                // if the watcher is dropped by the user without being disposed. If we can't get the watcher,
+                // there's nothing more to do (we can't raise events), so bail.
+                FileSystemWatcher watcher;
+                if (!_weakWatcher.TryGetTarget(out watcher))
                 {
-                    path += "/.";
+                    CancellationCallback();
```

Comment:
Is it possible this could get called multiple times?  If so, what happens in that case?